unit-tests to check ‘shortAddr’ function

DS-jr opened the pull request.

DS-jr committed:
unit-tests to check ‘shortAddr’ function

DS-jr commented:
@tectonick 
Please, check the error (appears during tests) in line N2 here:
https://github.com/ctrlsa/seal-app/commit/a666fd6f14ab6abe3dd8499d6d59ac8f84d450fc#diff-4d1c553a7ecea861c38acd7d95cd197a7c0e6dd1210d3c3197aa2d87a8715dc2R2

tectonick commented:
“@DS-jr, I think the root cause of this error is that the ﻿`shared/lib/utils` child import, ﻿`@twa-dev/sdk`, uses the ﻿window object itself during module initialization. Test environments like Jest or Vitest don’t include the ﻿window object from the browser API by default because tests are not run in a browser environment. However, this can be easily circumvented by adding additional packages that make the ﻿window object available in tests. For Vitest, it’s just a matter of setting the ﻿environment option to ﻿`jsdom` or ﻿`happy-dom` in the ﻿`vite.config.js` file. You can add the following configuration to your ﻿vite.config.js:
```
export default defineConfig({
  // other config options
  test: {
    environment: 'jsdom'
    // other test options 
  }
});
```

tectonick commented on tests/unit/index.test.js:
```diff
@@ -1,5 +1,27 @@
+// Unit-tests to verify "shortAddr(addr)" function accepts only a string as "addr" argument
+describe("shortAddr", () => {
+	it("Positive test cases - if 'addr' argument is a string", () => {
+		expect(shortAddr("string")).not.toThrow();
```
The ﻿`expect()` function, when used with the ﻿`toThrow()` matcher, requires a function as its argument. The current code, however, passes a string, which is the result of invoking a function, instead of passing the function itself. 
To correct this, you should enclose your function invocation within an arrow function or use the ﻿bind method to attach the parameter to the function. For example:
```
expect(() => shortAddr("234")).not.toThrow();
expect(shortAddr.bind(this, "234")).not.toThrow(); // This is a somewhat old-school approach
```

DS-jr replied:
Thanks! 
Done.

tectonick commented on tests/unit/index.test.js:
```diff
@@ -1,5 +1,27 @@
+		expect(() => shortAddr(true)).toThrow();
+		expect(() => shortAddr(null)).toThrow();
+		expect(() => shortAddr(undefined)).toThrow();
+		expect(() => shortAddr([])).toThrow();
```
Usually, it’s not necessary to test all JavaScript types. By the way, such errors are typically caught using TypeScript/JSDoc. I would suggest considering this approach in the future

DS-jr replied:
Got you, thanks.

tectonick replied:
I mean, it’s not really necessary here to test for every JavaScript type unless you expect them to be passed to your function at runtime.

And such JSDoc comment will solve most of such issues during development without event using TS explicitly (but it's TS under the hood):
```
/**
 * Shortens an address string to a more readable format by showing only the first 6 characters,
 * followed by an ellipsis, and then the last 4 characters.
 *
 * @param {string} addr - The address string to shorten.
 * @returns {string} The shortened address string.
 * @throws {TypeError} If the `addr` argument is not a string.
 */
export function shortAddr(addr) {
  if (typeof addr !== "string") {
    throw new TypeError("'addr' argument must be a string");
  }
  return addr.substring(0, 6) + "..." + addr.substring(addr.length - 4);
}
```
Check it in your editor. It might pick up such syntax by default and prevent you from accidentally passing incorrect types to the function. You can learn more here: https://dev.to/samuel-braun/boost-your-javascript-with-jsdoc-typing-3hb3

DS-jr committed:
fixed unit-test bugs

DS-jr commented:
@tectonick  
Fixed previous bugs with this commit: 
https://github.com/ctrlsa/seal-app/commit/acf8d52d5e56f9afd4ddd6bd5d5cf745ef96cf6a 

Unit-tests are working ok after that. 

If NO new bugs / mistakes in my code - please, confirm that https://github.com/ctrlsa/seal-app/pull/48 can be closed.

DS-jr commented:
@tectonick 
About this commit: 
​​https://github.com/ctrlsa/seal-app/commit/acf8d52d5e56f9afd4ddd6bd5d5cf745ef96cf6a 

I’ve changed manually only several lines of code (Ex.: added "jsdom" dependency), but 200+ modifications happened in the repo according to git.

This is how JS dependencies are updated automatically without my involvement, correct?

DS-jr committed:
fixing tests for 'isAuthenticated' function (in progress)

tectonick commented on tests/unit/index.test.js:
```diff
@@ -1,5 +1,49 @@
+	it("must return 'false' for an empty wallet", () => {
+		vi.mocked(get).mockReturnValueOnce({}); // TypeError: vi.mocked(...).mockReturnValueOnce is not a function
+		expect(isAuthenticated()).toBe(false);
+		// vi.mock(get, {}).mock(); // Error: Expression expected
```
I would suggest removing commented out code, it's not needed under source control.

tectonick commented on tests/unit/index.test.js:
```diff
@@ -1,5 +1,49 @@
+//   }
+describe("isAuthenticated", () => {
+	it("must return 'false' for an empty wallet", () => {
+		vi.mocked(get).mockReturnValueOnce({}); // TypeError: vi.mocked(...).mockReturnValueOnce is not a function
```
It's not expected during correct mocking to have such error `TypeError: vi.mocked(...).mockReturnValueOnce is not a function.`
The ﻿`vi.mocked` function allows TypeScript users to cast an object to a mocked type, although it does not offer any mocking functionality on its own.
Check this doc https://vitest.dev/guide/mocking.html, it shows good examples of correct mocking in vitest.

tectonick commented on tests/unit/index.test.js:
```diff
@@ -1,5 +1,49 @@
+
+
+// Unit-tests to verify "shortAddr(addr)" function accepts only a string as "addr" argument
+describe("shortAddr", () => {
```
Don't forget to include test cases for real wallet addresses, as they will be the most beneficial

tectonick commented on tests/unit/index.test.js:
```diff
@@ -1,5 +1,49 @@
-import { describe, expect, it } from "vitest";
+import { describe, expect, it, vi } from "vitest";
+import { shortAddr } from '/src/shared/lib/utils';
```
Using absolute paths like `/src/shared/lib/utils` in import statements is not standard and might not work as expected without additional configuration, since Node.js does not natively resolve such paths for modules.
It’s better to use relative paths to make sure your imports work as expected everywhere.

tectonick commented:
> @tectonick About this commit:​​[acf8d52](https://github.com/ctrlsa/seal-app/commit/acf8d52d5e56f9afd4ddd6bd5d5cf745ef96cf6a)
> 
> I’ve changed manually only several lines of code (Ex.: added "jsdom" dependency), but 200+ modifications happened in the repo according to git.
> 
> This is how JS dependencies are updated automatically without my involvement, correct?

@DS-jr it's ok. A package lock file is generated by package managers like npm, pnpm or yarn. It serves as a snapshot of the exact versions of all dependencies and their transitive dependencies installed in a project at a given time. Its changes should be committed to the git repository.

DS-jr committed:
fixing bugs after code review

DS-jr committed:
modified unit tests

DS-jr committed:
fixed tests for “changeTheme” function

DS-jr committed:
fixed tests for “shortAddr” function

DS-jr committed:
unit tests for “changeTheme(theme)” function: fixed again & improved

cmurav commented:
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
mArKmAnUeL1, @Mistybx, @Sim12courtney, @wanoo21, @HarryBO881, @yaiks, @ctaylor, @AbdelkaderBah, @tampham47, @00-Lightbarers-00, @hmo-5o1d, @thegoodresearch, @r0dreck, @jkurylowicz, @stargazersinternati, @melyalex, @DukeCaesar, @afiiijooon, @MELLVINI, @jlburke, @con44, @laveltonrcsw, @skaroskaro71, @adriancbo, @czarekphone, @tranvantien150295, @macrs, @nelin983, @asd0086, @choochoo1996, @MustardseedX, @TiegoEyes, @danBamikiya, @focusdada, @metalliccrunch, @krazygirl6, @NyanTunOo, @mcgingras, @Mlmolefe2, @sw82, @Jellyman447, @ks-hassam, @markmiro, @thnwatrthitiwitythnathnath, @paulpierre, @JennJenn1, @Belldoh, @Miminlaili, @TeaNicole023, @Drewire21

cmurav commented:
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
